### ISSUE TITLE
Foundry test Constructor can't deploy another contract

### ISSUE THREAD

**Issue by CodeSandwich** [matter-labs/foundry-zksync#794]
### Component

Other (please describe)

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

 forge 0.0.2 (1781234 2024-12-19T00:26:29.308590729Z)

### Operating System

Linux

#### 📝 Description

Using Foundry, contracts deployed in the test contract constructor lack the bytecode. If this turns out to be a generic zkSync compiler problem, I'll move this issue somewhere else.

#### 🔄 Reproduction Steps

Run `forge test --zksync --mc TargetTest -vvv` on a project containing this file:
```solidity
// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.20;

import "forge-std/Test.sol";

contract Target {
    uint256 public constant ZERO = 0;
}

contract TargetTest is Test {
    Target internal setUpTarget;
    Target internal immutable immutableTarget = new Target();
    Target internal immutable immutableConstructorTarget;
    Target internal storageTarget = new Target();
    Target internal storageConstructorTarget;

    constructor() {
        immutableConstructorTarget = new Target();
        storageConstructorTarget = new Target();
    }

    function setUp() public {
        setUpTarget = new Target();
    }

    function testLocal() public {
        callTarget(new Target());
    }

    function testSetUp() public view {
        callTarget(setUpTarget);
    }

    // FAIL
    function testImmutable() public view {
        callTarget(immutableTarget);
    }

    // FAIL
    function testConstructorImmutable() public view {
        callTarget(storageConstructorTarget);
    }

    // FAIL
    function testStorage() public view {
        callTarget(storageTarget);
    }

    // FAIL
    function testStorageConstructor() public view {
        callTarget(storageConstructorTarget);
    }

    function callTarget(Target target) internal view {
        console.log("Callng target", address(target));
        target.ZERO();
    }
}
```

#### 🤔 Expected Behavior

The tests pass.

#### 😯 Current Behavior

All tests relying on the constructor initialization are failing:
```
[FAIL: EvmError: Revert] testConstructorImmutable() (gas: 212578)
[FAIL: EvmError: Revert] testImmutable() (gas: 210457)
[FAIL: EvmError: Revert] testStorage() (gas: 212556)
[FAIL: EvmError: Revert] testStorageConstructor() (gas: 212555)
```
Each of the tests behaves very similarly, they cause an error:
```
2024-12-19T10:07:42.923390Z ERROR foundry_zksync_core::vm::tracers::cheatcode: call may fail or behave unexpectedly due to empty code target=0x5991a2df15a8f6a256d3ec51e99254cd3fb576a9 calldata="58fa63ca"
```
Followed by a test failure:
```
[FAIL: EvmError: Revert] testStorageConstructor() (gas: 212555)
Logs:
  Callng target 0x5991A2dF15A8F6A256D3Ec51E99254Cd3fb576A9

Traces:
  [212555] TargetTest::testStorageConstructor()
    ├─ [0] console::log("Callng target", Target: [0x5991A2dF15A8F6A256D3Ec51E99254Cd3fb576A9]) [staticcall]
    │   └─ ← [Stop] 
    ├─ [204154] Target::ZERO() [staticcall]
    │   └─ ← [Return] 
    └─ ← [Revert] EvmError: Revert
```
Considering that the `Target` is said to have no code the failure makes sense, it's a typical EOA call issue.

#### 🖥️ Environment

- Linux
- `forge 0.0.2 (1781234 2024-12-19T00:26:29.308590729Z)`
- `Solc 0.8.20`
- `zksolc-1.5.7`

#### 📋 Additional Context

No.

#### 📎 Log Output

No.

**Comment by Jrigada** [comment]
Hey @CodeSandwich, we are looking into this issue, in particular, the contract deployment in state variable initialization. For the contract deployment in the constructor, the tests are failing when trying to interact with the deployed contracts because the deployment occurs in the EVM context. At the same time, the subsequent call executes in EraVM. As a result, there is nothing at the address that was called.
This issue can be circumvented using the intended cheatcode vmExt.zkVm(true), which forces the execution to switch from EVM to EraVM and performs the deployments there accordingly.
```
    constructor() {
	vmExt.zkVm(true);
        Target immutableConstructorTarget = new Target();
        Target storageConstructorTarget = new Target();
    }
```
Also, moving the contract deployment in the state variables to the constructor would solve the other deployments as well.
I will look into it in the following days with the team. Thank you for the input.

**Comment by CodeSandwich** [comment]
That's interesting, what are the EVM and EraVM contexts? Why would I select one or another? What's the default behavior, the test contracts are deployed locally in the EVM context, are other contracts and other environments treated like this as well? Why is the default like this and what am I risking or losing by defying it?

**Comment by Jrigada** [comment]
Good questions, in Foundry-zkSync, tests initially execute in the EVM context (the traditional execution environment for Ethereum smart contracts). This is done to maintain compatibility with Ethereum tooling and to leverage Foundry features like cheatcodes.

To test zkSync-specific features, such as Paymaster, Account Abstraction (AA), or to target the zkSync network, you can switch to the EraVM context (zkSync Era's custom virtual machine). This can be achieved using either the `--zksync` flag or the `vm.zkVm(true)` cheatcode. By doing so, contract deployments and function calls are executed within the zkSync environment, ensuring alignment with its execution model.

For a more detailed explanation, refer to the [Execution Overview](https://foundry-book.zksync.io/zksync-specifics/execution-overview) in the Foundry-zkSync documentation.

In summary, a forge test begins execution in the EVM context but can switch to the EraVM context during the test for zkSync-specific features. Selecting the EraVM context ensures that all calls and deployments are executed within zkSync.

**Comment by CodeSandwich** [comment]
Is this complete as in "the constructor can deploy contracts now" or as in "the user who opened the issue was told to not deploy contracts in the constructor"?

**Comment by CodeSandwich** [comment]
Checked again with `forge 0.0.2 (cb40c35 2025-01-07T00:26:34.982764333Z)` and it's still broken. Why do you not want to fix it?
